apps/prairielearn/python/prairielearn/timeouts.py

```diff
@@ -0,0 +1,161 @@
+        self.state = TimeoutState.EXECUTED
+
+    def __bool__(self) -> bool:
+        """Return whether the context is not TIMED_OUT or INTERRUPTED"""
+        return self.state in (
+            TimeoutState.EXECUTED,
+            TimeoutState.EXECUTING,
+            TimeoutState.CANCELED,
+        )
+
+    def __repr__(self) -> str:
+        """Debug helper"""
+        return f"<{self.__class__.__name__} in state: {self.state}>"
+
+    def __enter__(self) -> Self:
+        """Initializes the interrupt and updates state."""
+        self.state = TimeoutState.EXECUTING
+        self.setup_interrupt()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> bool:
+        """Manages exceptions & timeout."""
+        if exc_type is TimeoutExceptionError:
+            if self.state != TimeoutState.TIMED_OUT:
+                self.state = TimeoutState.INTERRUPTED
+                self.suppress_interrupt()
+            return self.swallow_exc
+        else:
+            if exc_type is None:
+                self.state = TimeoutState.EXECUTED
+            self.suppress_interrupt()
+        return False
+
+    def cancel(self) -> None:
+        """In case in the block you realize you don't need anymore
+        limitation
+        """
+        self.state = TimeoutState.CANCELED
+        self.suppress_interrupt()
+
+    # Methods must be provided by subclasses
+    def suppress_interrupt(self) -> None:
+        """Removes/neutralizes the feature that interrupts the executed block"""
+        raise NotImplementedError
+
+    def setup_interrupt(self) -> None:
+        """Installs/initializes the feature that interrupts the executed block"""
+        raise NotImplementedError
+
+
+def async_raise(target_tid: int, exception: type[Exception]) -> None:
+    """Raises an asynchronous exception in another thread.
+    Read http://docs.python.org/c-api/init.html#PyThreadState_SetAsyncExc
+    for further enlightenments.
+
+    :param target_tid: target thread identifier
+    :param exception: Exception class to be raised in that thread
+
+    Raises:
+        ValueError: If the thread ID is invalid.
+        SystemError: If the process fails to set AsyncExec
+    """
+    # TODO: Update this code when we are running no-GIL Python version 3.14+.
+
+    # Ensuring and releasing GIL are useless since we're not in C
+    # gil_state = ctypes.pythonapi.PyGILState_Ensure()
+    ret = ctypes.pythonapi.PyThreadState_SetAsyncExc(
+        tid_ctype(target_tid), ctypes.py_object(exception)
+    )
+    # ctypes.pythonapi.PyGILState_Release(gil_state)
+    if ret == 0:
+        raise ValueError(f"Invalid thread ID {target_tid}")
+    if ret > 1:
+        ctypes.pythonapi.PyThreadState_SetAsyncExc(tid_ctype(target_tid), None)
+        raise SystemError("PyThreadState_SetAsyncExc failed")
+
+
+class ThreadingTimeout(BaseTimeout):
```

Comment:
Can we condense these two classes? As used here, this class just overrides some methods not defined in the base class (i.e. everything can just be in one class).